cmd/internal/cli/actions_linux.go

```diff
@@ -46,76 +49,149 @@ import (
+	}
+
+	if err2 := os.RemoveAll(tempDir); err2 != nil {
+		sylog.Errorf("Couldn't remove temporary directory %s: %s", tempDir, err2)
+	}
+	return false, "", "", fmt.Errorf("while extracting image: %w", err)
+}
+
+// extractImage extracts img to directory dir within a temporary directory
+// tempDir. It is the caller's responsibility to remove tempDir
+// when no longer needed.
+func extractImage(img *imgutil.Image, imageDir string) error {
+	sylog.Infof("Converting SIF file to temporary sandbox...")
+	unsquashfsPath, err := bin.FindBin("unsquashfs")
+	if err != nil {
+		return err
 	}
 
 	// create a reader for rootfs partition
 	reader, err := imgutil.NewPartitionReader(img, "", 0)
 	if err != nil {
-		return "", "", fmt.Errorf("could not extract root filesystem: %s", err)
+		return fmt.Errorf("could not extract root filesystem: %s", err)
 	}
 	s := unpacker.NewSquashfs()
 	if !s.HasUnsquashfs() && unsquashfsPath != "" {
 		s.UnsquashfsPath = unsquashfsPath
 	}
 
-	// keep compatibility with v2
-	tmpdir := os.Getenv("SINGULARITY_TMPDIR")
-	if tmpdir == "" {
-		tmpdir = os.Getenv("SINGULARITY_LOCALCACHEDIR")
-		if tmpdir == "" {
-			tmpdir = os.Getenv("SINGULARITY_CACHEDIR")
-		}
+	// extract root filesystem
+	if err := s.ExtractAll(reader, imageDir); err != nil {
+		return fmt.Errorf("root filesystem extraction failed: %s", err)
 	}
 
-	// create temporary sandbox
-	tempDir, err = ioutil.TempDir(tmpdir, "rootfs-")
+	return nil
+}
+
+// squashfuseMount mounts img using squashfuse to directory imageDir. It is the
+// caller's responsibility to umount imageDir when no longer needed.
+func squashfuseMount(img *imgutil.Image, imageDir string) (err error) {
+	part, err := img.GetRootFsPartition()
 	if err != nil {
-		return "", "", fmt.Errorf("could not create temporary sandbox: %s", err)
+		return fmt.Errorf("while getting root filesystem : %s", err)
 	}
-	defer func() {
-		if err != nil {
-			os.RemoveAll(tempDir)
-		}
-	}()
-
-	// create an inner dir to extract to, so we don't clobber the secure permissions on the tmpDir.
-	imageDir = filepath.Join(tempDir, "root")
-	if err := os.Mkdir(imageDir, 0o755); err != nil {
-		return "", "", fmt.Errorf("could not create root directory: %s", err)
+	if img.Type != image.SIF && part.Type != image.SQUASHFS {
+		return fmt.Errorf("only SIF images are supported")
+	}
+	if IsFakeroot {
+		return fmt.Errorf("fakeroot is not currently supported")
 	}
+	sylog.Infof("Mounting SIF with FUSE...")
 
-	// extract root filesystem
-	if err := s.ExtractAll(reader, imageDir); err != nil {
-		return "", "", fmt.Errorf("root filesystem extraction failed: %s", err)
+	squashfusePath, err := bin.FindBin("squashfuse")
+	if err != nil {
+		return fmt.Errorf("squashfuse is required: %w", err)
+	}
+	if _, err := bin.FindBin("fusermount"); err != nil {
+		return fmt.Errorf("fusermount is required: %w", err)
 	}
 
-	return tempDir, imageDir, err
+	return sifuser.Mount(context.TODO(), img.Path, imageDir,
```

Comment:
Would be good to wire this further up the chain, perhaps `cobraCmd.Context()` in `execStarter`?